Recursive as dictionary

amarkpayne opened the pull request:
### Motivation or Problem
The methods `as_dict` and `make_object` allow for any object that inherits from RMGObject to be easily saved to a yaml file and reconstructed from a yaml file, respectively. However, as currently written is RMG there is a limited recursion depth to how many objects, dictionaries, and lists can be nested inside of each other.

For example, consider an object that has an attribute that is a dictionary which itself contains another object. The desired affect of `as_dict` would be for the object nested inside of this dictionary to be expanded to a dictionary (i.e. for `as_dict` to be called on this object as well). However, the current implementation will return for this attribute a dictionary that contains the object itself rather than the expanded object.

Furthermore, many objects have to rewrite their `as_dict` and `make_objects` to function properly, even if they inherit from RMGObject. Ideally, newly created objects that inherit from RMGObject should have working `as_dict` and `make_object` methods without needing to overwrite these methods (at least more times than not).

### Description of Changes
Inside the file where we define the RMGObject I have created method for recursively calling `as_dict` for all objects inside of a dictionary or list. I have also created a method for recursively calling `make_object` for all objects inside of a dictionary. Becasue these functions are recursive by calling themselves, the recursion limit for nested objects and iterables should be whatever the recursion stack limit of python is (very large). 

While these functions can be called as standalone functions (potentially useful for scripting), I have used these functions to help re-write the `as_dict` and `make_object` methods of the RMGObject class.

With this, I have also cleaned-up the affected code, and made it so that the `as_dict` and `make_object` methods are only re-written when absolutely necessary (for example, see statmech rotations and torsions), which thanks to the new functions require very little code even when necessary. Note that this is usually necessary whenever passing all attributes to the __init__ method causes issues.

### Testing
I have ran the unit tests, and even successfully recreated an ArkaneSpecies object from a YAML file, but it is likely that there are affected classes that weren't properly unit tested. I am specifically concerned about the Wilhoit class, and the NASA class can probably be cleaned up further.

### Reviewer Note
Because we have a few pull requests open or about to be opened that define new objects that will be saved to a YAML file, these PRs will require these changes, and thus this PR needs to be merged in first. For example, this affects #1577, #1641, and #1561

alongd commented on rmgpy/rmgobject.pyx:
```diff
@@ -115,3 +115,21 @@ cdef class RMGObject(object):
+
+    elif isinstance(obj, dict):
+        for key, value in obj.iteritems():
+            obj[key] = expand_to_dictionaries(value)
```
what if the key is an object as well?

alongd replied:
Are we sure this won't change the original object?

amarkpayne replied:
Just to document this here, we figured out that dictionaries themselves are not hashable, and thus cannot serve as keys for other dictionaries. Therefore, we can not simply expand objects to dictionaries if they are being used as keys for a dictionary. 

Max and I had some ideas for how we might tackle this, but this is a subject for a later PR due to its complexity and the fact that no RMGObjects currently have attributes that are dictionaries with objects as keys (that is until we make Molecule an RMGObject)

alongd commented on rmgpy/rmgobject.pyx:
```diff
@@ -115,3 +115,21 @@ cdef class RMGObject(object):
                     pass
         self.__init__(**data)
+
+cpdef expand_to_dictionaries(obj):
```
Maybe consider changing the name, since the function deals with more than just dicts

alongd replied:
Add a docstring?

mliu49 replied:
I like shorter names as long as they're still descriptive. Maybe just "expand_to_dict" or "repr_dict"?

mliu49 replied:
Another question, would there be any benefit to making this and `recursive_make_object` class methods of RMGObject?

amarkpayne replied:
I'll refactor to `expand_to_dict`.

amarkpayne replied:
As @mliu49 and I discussed offline, there would be benefit to making `expand_to_dict` and `recursive_make_object` static (as opposed to class) method of RMGObject, as we could call these functions from any class that inherits from RMGObject (meaning we would not have to import these functions explicitly most of the time). The downside of this is that these functions work on (hopefully) all objects, not just those that subclass RMGObject. Therefore, to avoid this confusion we will keep this as is

alongd commented on rmgpy/rmgobject.pyx:
```diff
@@ -115,3 +115,21 @@ cdef class RMGObject(object):
+        return obj
+
+    elif isinstance(obj, np.ndarray):
+        return obj.tolist()
```
Do we also know to read a list and generate a numpy array if needed?

alongd commented on rmgpy/rmgobject.pyx:
```diff
@@ -133,3 +134,52 @@ cpdef expand_to_dictionaries(obj):
+                                                                                                class_dictionary))
+
+            for key, val in obj.iteritems():
+                if key == 'class':
```
maybe change `==` to `!=` and skip the `continue`?

amarkpayne replied:
Thanks for the catch!

alongd commented on rmgpy/rmgobject.pyx:
```diff
@@ -133,3 +134,52 @@ cpdef expand_to_dictionaries(obj):
         else:
             return obj
+
+cpdef recursive_make_object(obj, class_dictionary, make_final_object=True):
```
Do we ever call this with `make_final_object=False`?

amarkpayne replied:
Yes, this gets called with `make_final_object=False` in the RMGObject `make_object` method. This is because the `make_object` method passes the dictionary as keyword arguments into the __init__ method. By doing this, it initializes the object in place.

If you call the function with `make_final_object=True` then the function returns a new instance of the object (i.e. not in place). This mean that if you read in a dictionary from a yaml file, you can simply call this function on the dictionary and it will return the object for you (you do not need to create an "empty" instance of the object first.

However, if you already have an empty instance or want to reinitialize the object, you just need to call the `make_object` method of the object.

alongd commented on rmgpy/rmgobject.pyx:
```diff
@@ -133,3 +134,52 @@ cpdef expand_to_dictionaries(obj):
+    elif isinstance(obj, list):
+        return [recursive_make_object(x, class_dictionary) for x in obj]
+
+    elif isinstance(obj, str):
```
Can you do `isinstance(obj, (str, unicode))` instead? In Py3 (and Py-3 compatible code like ARC) all strings are unicode by default

amarkpayne replied:
Done, thanks for the suggestion!

mliu49 replied:
In Py3, there is no unicode class, rather the default str class is unicode. So when we transition, I think we would need to remove the check for unicode? Is this correct?

alongd commented on rmgpy/rmgobject.pyx:
```diff
@@ -133,3 +134,52 @@ cpdef expand_to_dictionaries(obj):
+
+    elif isinstance(obj, str):
+        # Check to see if the string needs to be converted to a float or int
+        if bool(re.search('\d', obj)):
```
Maybe search for `^\d+$` (start to end only digits) instead of the try except

amarkpayne replied:
I don't think this works for floats (try `bool(re.search('^\d+$', '12.45'))` in python 2). I'll keep looking for a better method though.

alongd replied:
will `^[\d.]+$` work?

amarkpayne replied:
Closer, but it doesn't catch floats in scientific notation like `'-1.20e-05'` and negative numbers

mliu49 replied:
Idk if this is a good solution, but could you just try casting to float, then int if that fails, then leave as string if that fails?

amarkpayne replied:
Is there an instance where something could be cast as `int` but not `float`? Also, would this convert all `int`s to `float`s?

mliu49 replied:
Oh yeah, that's true. Is there a benefit to the first search for digits? Could you skip directly to the check for a decimal point?

alongd commented on arkane/common.py:
```diff
@@ -252,7 +252,8 @@ def load_yaml(self, path, species, pdep=False):
         if class_name != 'ArkaneSpecies':
             raise KeyError("Expected a ArkaneSpecies object, but got {0}".format(class_name))
-        del data['class']
+        data = {key: data[key] for key in data.iterkeys() if key != 'class'}
```
What's the difference between this and the deleted line?

amarkpayne replied:
The deleted line deletes class from the original object, which changes data. In this file I realize now that this isn't an issue, and I can revert this back (it might be faster to use `del data['class']`).

I had used code like this in other files, where changing `data` could cause problems under certain circumstances. For example, if you were scripting and had defined `data` externally but passed it to functions like `recursive_make_dictionaries` multiple times.

Let me know what your preference on this is, I am happy to leave it in or revert this line.

alongd replied:
I think the original like is simpler. If it's not problematic any more, lets revert the change.
Same for `imaginary_frequency` below?

alongd commented on arkane/common.py:
```diff
@@ -268,19 +269,19 @@ def load_yaml(self, path, species, pdep=False):
             freq_data = data['imaginary_frequency']
-            del data['imaginary_frequency']
+            data = {key: data[key] for key in data.iterkeys() if key != 'imaginary_frequency'}
+        data['species'] = Species().fromSMILES(data['smiles'])
```
sometimes an adjlist or InChI are given, and sometimes non of them

amarkpayne replied:
I think this is something I have been struggling with a little bit. It is hard because we need an rmgpy Species to create an ArkaneSpecies, but because we don't have a species attribute on ArkaneSpecies the species object does not get saved to the yaml file. How should this work ideally? How can we reconstruct an ArkaneSpecies from a yaml file if the file does not have smiles or InChI?

amarkpayne replied:
Settled by allowing a species to be created from SMILES, InChI, or adjacency list

alongd commented on arkane/common.py:
```diff
@@ -268,19 +269,19 @@ def load_yaml(self, path, species, pdep=False):
-        self.adjacency_list = data['adjacency_list'] if 'adjacency_list' in data else None
-        self.inchi = data['inchi'] if 'inchi' in data else None
-        self.smiles = data['smiles'] if 'smiles' in data else None
-        self.is_ts = data['is_ts'] if 'is_ts' in data else False
```
Are these no longer necessary?

amarkpayne replied:
I don't believe these are necessary, but I could be wrong. I tested generating ArkaneSpecies yaml files, and everything looked okay (but I'll check again).

I believe this is the case because if these attributes were in `data` then these attributes would be passed to ArkaneSpecies.__init__() by `make_object`, and all 4 of these can be sent to __init__() and will be set there.

If there is a more subtle reason why these things were there though it is possible that they may still be needed.

alongd replied:
I think these are here just to report information in the YAML file, since Species doesn't have a .smiles .adjlist, .inchi, and .is_ts attributes. They shouldn't be processed when reading a YAML file, though. I think it's nice and convenient to have adjList / SMILES in the YAML file

amarkpayne replied:
I am not sure I follow completely--let's talk offline about this

amarkpayne replied:
@alongd and I discussed this offline. We ran an example Arkane job to confirm that these attributes are indeed stored on the YAML file even without these lines as expected.

rwest commented:
Not had chance to review the code, but from the pull request description this sounds like a great improvement. 👏🏻

amarkpayne commented:
Thanks @alongd for the timely review and great comments! I have addressed your comments and added fixup commits to add changes where I could. Let me know what you think. In the meantime I think it might be worthwhile if I add some additional unit testing to catch some of the points that you addressed.

amarkpayne commented:
@alongd I have applied a few of the changes that we discussed offline and also added some unit tests. I also further refactored quantity, NASA/NASAPolynomial, and Wilhoit. With this I have added everything that I wanted to add for this PR, so this should be ready for one last round of review (and final changes if necessary).

amarkpayne commented:
As we discussed offline, the fixup commits that I squashed are on the branch `rad_w_fixup` on official

alongd commented on arkane/common.py:
```diff
@@ -274,15 +275,23 @@ def load_yaml(self, path, species, pdep=False):
+            if 'smiles' in data:
+                data['species'] = Species().fromSMILES(data['smiles'])
+            elif 'adjacency_list' in data:
+                data['species'] = Species().fromAdjacencyList(data['adjacency_list'])
```
would you like to add inchi as well? Maybe using something like:
`Species(molecule=[Molecule().fromInChI().generate_resonance_structures()])`
or is there a simpler method?

mliu49 replied:
You can now do `Species(InChI='whatever')` and `Species(SMILES='whatever')` for instantiation.

amarkpayne replied:
Thanks @mliu49 ! I'll implement this then

alongd commented on rmgpy/quantity.py:
```diff
@@ -227,19 +228,6 @@ def as_dict(self):
             output_dict['uncertaintyType'] = self.uncertaintyType
         return output_dict
-
-    def make_object(self, data, class_dict):
```
Great!

alongd commented on rmgpy/statmech/rotation.pyx:
```diff
@@ -83,6 +84,12 @@ cdef class Rotation(Mode):
         """
         return (Rotation, (self.symmetry, self.quantum))
 
+    def make_object(self, data, class_dict):
```
Add in rotation.pxd as well, or is the definition already there?

(Same for torsion)

amarkpayne replied:
Good question, a thought the same when I first did this. Because `Rotation` inherits from `RMGObject`, this method is already present in this class (and was declared in rmgobject.pxd) and does not need to be declared in the header. All we are doing here is overwriting the parent method.

I am pretty sure I tried adding this to the header file initially, but I got an error. I'll check this one more time though.

mliu49 commented on rmgpy/rmgobject.pyx:
```diff
@@ -115,3 +115,35 @@ cdef class RMGObject(object):
+    of the object (useful for __repr__ methods and outputting to YAML). The function works recursively to all objects
+    nested in the original object.
+    
+    :param obj: Object to be represented by a dictionary (any type)
```
Yeah, I like the Google style docstring a bit more. I don't think we have an "official" stance though.

amarkpayne replied:
I'll switch to this then. I also figured out how to make pycharm default to this style here: https://www.jetbrains.com/help/pycharm/settings-tools-python-integrated-tools.html

amarkpayne commented on rmgpy/rmgobject.pyx:
```diff
@@ -147,3 +148,69 @@ cpdef expand_to_dictionaries(obj):
+            args = {key:new_obj[key] for key in new_obj if key != 'class'}
+            if make_final_object:
+                if hasattr(class_to_make, 'make_object'):
+                    created_obj = class_to_make.__new__(class_to_make)
```
Nevermind, this causes more problems than it is worth, i'll stick with `if hasattr(class_to_make, 'make_object'):`

mliu49 commented on rmgpy/rmgobject.pyx:
```diff
@@ -53,65 +58,127 @@ cdef class RMGObject(object):
+        for key, value in obj.iteritems():
+            new_key = expand_to_dictionaries(key)
+            new_value = expand_to_dictionaries(value)
+            new_obj[new_key] = new_value
```
Can you use dictionaries as dictionary keys?

amarkpayne commented:
I have added the last of the changes that @mliu49 and I discussed here and offline. @alongd let me know if there are any other changes. Otherwise if you approve I will update the branch and squash the fixup commits in preparation for merging.

amarkpayne commented:
I added your formatting changes and squashed all of the fixup commits. Thanks again @mliu49 !

amarkpayne commented:
The builds are getting errors related to cairo. It looks like this package was updated on the RMG channel today. @mliu49 any idea if something went wrong with the new build?

mliu49 commented:
Sorry, I was playing around with updating a couple dependencies yesterday. The newest noarch cairocffi build by conda-forge apparently works on mac but not linux. The linux-64 build works though, so I just removed the noarch build.

alongd commented on arkane/common.py:
```diff
@@ -231,21 +231,26 @@ def save_yaml(self, path):
-        except KeyError:
-            # Lacking label in the YAML file is strange, but accepted
-            logging.debug('Did not find label for species {0} in .yml file.'.format(species.label))
+        if label:
```
You can move it to the `if label` condition above

amarkpayne replied:
I thought about that, but then the "Loading ..." message will have to come after we load `data = yaml.safe_load(stream=f)`. Which would you prefer?

alongd commented on arkane/common.py:
```diff
@@ -231,21 +231,26 @@ def save_yaml(self, path):
         """
-        logging.info('Loading statistical mechanics parameters for {0} from .yml file...'.format(species.label))
+        yml_file = os.path.basename(path)
+        if label:
```
thanks!!

alongd commented on arkane/common.py:
```diff
@@ -270,19 +275,33 @@ def load_yaml(self, path, species, pdep=False):
+                raise ValueError('Cannot load ArkaneSpecies from YAML file {0}. Either `smiles`, `adjacency_list`, or '
+                                 'InChI must be specified'.format(path))
+            # Finally, set the species label so that the special attributes are updated properly
+            if label:  # By default use the label passed by the user
```
for clarity, maybe relocate the label setting near the `if label != data['label']:` block?

amarkpayne replied:
Done

alongd commented on rmgpy/rmgobject.pyx:
```diff
@@ -42,6 +41,10 @@ cdef class RMGObject(object):
+        
+        Returns: 
+            dict: A dictionary representation of the object
+
```
remove extra line break?

amarkpayne replied:
Done

alongd commented on rmgpy/rmgobject.pyx:
```diff
@@ -53,68 +56,24 @@ cdef class RMGObject(object):
+
+        Returns: 
+            None
+
```
remove extra line break?

amarkpayne replied:
Done

alongd commented on arkane/commonTest.py:
```diff
@@ -334,25 +335,67 @@ def test_dump_yaml(self):
+
+        # Load in newly created YAML file
+        arkane_spc_old = jobList[0].arkane_species
+        arkane_spc = ArkaneSpecies.__new__(ArkaneSpecies)
```
What's the difference between this and `arkane_spc = ArkaneSpecies()`?

amarkpayne replied:
I believe that `arkane_spc = ArkaneSpecies()` raises an error, because ArkaneSpecies objects need to be instantiated either with an rmgpy species object or a conformer object. We need an object in order to call the `load_yaml` method, but we don't want to try and instantiate the object until we call `load_yaml`

amarkpayne replied:
Confirmed by the following:

```
>>> from arkane.common import ArkaneSpecies
>>> arkane_spc = ArkaneSpecies()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "RMG-Py/arkane/common.py", line 95, in __init__
    raise ValueError('No species (or TS) or conformer was passed to the ArkaneSpecies object')
ValueError: No species (or TS) or conformer was passed to the ArkaneSpecies object
```

amarkpayne commented:
The last round of fixes/additions has been rebased/autosquashed. Let me know if this is ready to merge or if there are any other changes we missed.

amarkpayne commented:
Thanks @alongd and @mliu49 for all of the hard work you put into this PR!
